packages/flutter_tools/lib/src/build_system/phases.dart

```diff
@@ -0,0 +1,180 @@
+///
+/// A [BuildDefinition] can be executed via `flutter assemble`.
+class BuildDefinition {
+  /// Create a new [BuildDefinition].
+  ///
+  /// Both [name] and [phases] must not be null. Phases must not be empty
+  /// or [createBuild] will throw a [StateError].
+  const BuildDefinition({
+    @required this.name,
+    @required this.phases,
+  }) : assert(name != null && name != ''),
+       assert(phases != null);
+
+  final String name;
+  final List<BuildPhase> phases;
+
+  /// Create the root [Target] to be invoked via the [BuildSystem].
+  ///
+  /// This method is also responsible for validating the target graph, including
+  /// checking for cycles and naming collisions.
+  ///
+  /// Throws a [StateError] if [phases] is empty.
+  // TODO(jonahwilliams): add output file collision detection.
+  Future<Target> createBuild(Environment environment) async {
+    if (phases.isEmpty) {
+      throw StateError('A BuildDefinition was created with no phases.');
+    }
+    // First create an aggregate target for each phase and store then by.
+    // phase name.
+    final Map<String, _AggregateTarget> aggregates = <String, _AggregateTarget>{};
+    for (BuildPhase buildPhase in phases) {
+      final List<Target> targets = await buildPhase.plan(environment);
+      final _AggregateTarget target = _AggregateTarget(targets, buildPhase);
+      aggregates[target.name] = target;
+    }
+    // Then order the aggregate targets such that dependencies are inserted into
+    // the target graph in order. This is by traversing the list of aggregate
+    // nodes and inserting the dependant aggregate targets into the list of
+    // dependencies. This is validated by the call to [checkCycles] later.
+    final _RootTarget result = _RootTarget(aggregates.values.toList(), this);
+    for (_AggregateTarget target in aggregates.values) {
+      for (String phaseDependency in target.phase.dependencies) {
+        final _AggregateTarget dependency = aggregates[phaseDependency];
+        target.dependencies.add(dependency);
+      }
+    }
+    checkCycles(result);
+    return result;
+  }
+}
+
+// A synthetic target created from a List of [BuildPhase] aggregate targets.
+class _RootTarget extends Target {
+  _RootTarget(this.dependencies, this.definition);
+
+  final BuildDefinition definition;
+
+  @override
+  final List<Target> dependencies;
+
+  @override
+  String get name => definition.name;
+
+  @override
+  Future<void> build(List<File> inputFiles, Environment environment) async { }
+
+  @override
+  List<Source> get inputs => const <Source>[];
+
+  @override
+  List<Source> get outputs => const <Source>[];
+}
+
+// A synthetic target created from the result of a [BuildPhase.plan].
+class _AggregateTarget extends Target {
+  _AggregateTarget(this.dependencies, this.phase);
+
+  final BuildPhase phase;
+
+  @override
+  Future<void> build(List<File> inputFiles, Environment environment) async { }
+
+  @override
+  final List<Target> dependencies;
```

Comment:
Currently the target API uses a list of instances, since this tries to represent the actual build graph as a graph, as opposed to the buildPhase which is just a recipe for construction of a build. Put another way, if all I had were strings here, I'd need a third structure to store the associations when I'm walking the build graph.

Reply:
The logic to detect cycles seems to assume that the target is unique. In this case, do we enforce the uniqueness by looking up a map or set? A map could use the target's name to get the corresponding `Target`.  The main advantage of using strings is that we can chain targets without having to pass the references around.

Reply:
You're right, there is actually a requirement that the name is unique (per instance). I've updated the cycle detection logic to handle the case where there are multiple targets with the same name that aren't identical.